Sources/Basic/Path.swift

```diff
@@ -0,0 +1,473 @@
+        
+        // Create an output buffer using the capacity we've calculated.
+        // FIXME: Determine whether this is the most efficient way to do it.
+        var result = ""
+        result.reserveCapacity(capacity)
+        
+        // Put the normalized parts back together again.
+        var iter = parts.makeIterator()
+        result.append(PathSeparatorCharacter)
+        if let first = iter.next() {
+            result.append(contentsOf: first)
+            while let next = iter.next() {
+                result.append(PathSeparatorCharacter)
+                result.append(contentsOf: next)
+            }
+        }
+        
+        // Sanity-check the result (including the capacity we reserved).
+        assert(!result.isEmpty, "unexpected empty string")
+        assert(result.characters.count == capacity, "count: " +
+            "\(result.characters.count), cap: \(capacity)")
+        
+        // Use the result as our stored string.
+        _impl = PathImpl(string: String(result))
+    }
+    
+    /// Directory component.  An absolute path always has a non-empty directory
+    /// component (the directory component of the root path is the root itself).
+    public var dirname: String {
+        return _impl.dirname!
+    }
+    
+    /// Last path component (including the suffix, if any).  it is never empty.
+    public var basename: String {
+        return _impl.basename
+    }
+    
+    /// Suffix (including leading `.` character) if any.  Note that a basename
+    /// that starts with a `.` character is not considered a suffix.
+    public var suffix: String? {
+        return _impl.suffix
+    }
+    
+    /// Absolute path of parent directory.  This always returns a path, because
+    /// every directory has a parent (the parent directory of the root directory
+    /// is considered to be the root directory itself).
+    public var parentDirectory: AbsolutePath {
+        return isRoot ? self : AbsolutePath(_impl.dirname!)
+    }
+    
+    /// True if the path is the root directory.
+    public var isRoot: Bool {
+        let chars = _impl.string.characters
+        return chars.count == 1 && chars.first == PathSeparatorCharacter
+    }
+    
+    /// Returns the absolute path with the relative path applied.
+    public func join(_ subpath: RelativePath) -> AbsolutePath {
+        let slash = String(PathSeparatorCharacter)
+        return AbsolutePath(asString + slash + subpath.asString)
+    }
+    
+    /// NOTE: We will want to add other methods here, such as a join() method
+    ///       that takes an arbitrary number of parameters, etc.  Most likely
+    ///       we will also make the `+` operator mean `join()`.
+    
+    /// Root directory (whose string representation is just a path separator).
+    public static var root: AbsolutePath {
+        // FIXME: If this is called a lot, we should maybe cache the root path.
+        return AbsolutePath(String(PathSeparatorCharacter))
```

Comment:
It isn't free, because of thread safety (or can this be created statically, because it's a value type? — if so then I agree we should just do it, though I don't know how often this will get called).